moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+            val conflicting = get(toAdapters, toAdapter.type, toAdapter.annotations)
+            if (conflicting != null) {
+              throw IllegalArgumentException(
+                """Conflicting @ToJson methods:
+    ${conflicting.method}
+    ${toAdapter.method}"""
+              )
+            }
+            toAdapters.add(toAdapter)
+          }
+          if (m.isAnnotationPresent(FromJson::class.java)) {
+            val fromAdapter = fromAdapter(adapter, m)
+            val conflicting = get(fromAdapters, fromAdapter.type, fromAdapter.annotations)
+            if (conflicting != null) {
+              throw IllegalArgumentException(
+                """Conflicting @FromJson methods:
+    ${conflicting.method}
+    ${fromAdapter.method}"""
+              )
+            }
+            fromAdapters.add(fromAdapter)
+          }
+        }
+        clazz = clazz.superclass
+      }
+      if (toAdapters.isEmpty() && fromAdapters.isEmpty()) {
+        throw IllegalArgumentException("Expected at least one @ToJson or @FromJson method on ${adapter.javaClass.name}")
+      }
+      return AdapterMethodsFactory(toAdapters, fromAdapters)
+    }
+
+    /**
+     * Returns an object that calls a `method` method on `adapter` in service of
+     * converting an object to JSON.
+     */
+    fun toAdapter(adapter: Any, method: Method): AdapterMethod {
+      method.isAccessible = true
+      val returnType = method.genericReturnType
+      val parameterTypes = method.genericParameterTypes
+      val parameterAnnotations = method.parameterAnnotations
+      return if (parameterTypes.size >= 2 &&
+        parameterTypes[0] == JsonWriter::class.java &&
+        returnType == Void.TYPE &&
+        parametersAreJsonAdapters(2, parameterTypes)
+      ) {
+        // void pointToJson(JsonWriter jsonWriter, Point point) {
+        // void pointToJson(JsonWriter jsonWriter, Point point, JsonAdapter<?> adapter, ...) {
+        val qualifierAnnotations = parameterAnnotations[1].jsonAnnotations
+        object : AdapterMethod(
+          parameterTypes[1],
+          qualifierAnnotations,
+          adapter,
+          method,
+          parameterTypes.size,
+          2,
+          true
+        ) {
+          override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+            invoke(writer, value)
+          }
+        }
+      } else if (parameterTypes.size == 1 && returnType != Void.TYPE) {
+        // List<Integer> pointToJson(Point point) {
+        val returnTypeAnnotations = method.jsonAnnotations
+        val qualifierAnnotations = parameterAnnotations[0].jsonAnnotations
+        val nullable = parameterAnnotations[0].hasNullable
+        object : AdapterMethod(
+          parameterTypes[0],
+          qualifierAnnotations,
+          adapter,
+          method,
+          parameterTypes.size,
+          1,
+          nullable
+        ) {
+          private var delegate: JsonAdapter<Any>? = null
```

Comment:
use `by lazy` rather than the manual lazy init?

Reply:
https://github.com/square/moshi/pull/1467/commits/3e6e9bda3372997dff18ac5d1567f6b80707074e#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R267